File: lab02/template method/SixStepTemplateMethod.cpp
/*---------------------------------------------------------------------------*\
| \file     SixStepTemplateMethod.cpp
| \brief    Six Step Template Method pattern implementation
| \author   Norm Evangelista
\*---------------------------------------------------------------------------*/
#include "SixStepTemplateMethod.hpp"

#include <cstdlib>
#include <iostream>
#include <random>

namespace template_method_common
{
/*---------------------------------------------------------------------------*\
|   SixStepTemplateMethod base class methods
\*---------------------------------------------------------------------------*/
SixStepTemplateMethod::SixStepTemplateMethod(
    const char* name = "" )
    : m_implementation_name( name )
{
    std::random_device                  rd;
    std::uniform_int_distribution<int>  uni( 0, 23 );

    auto random_hour = uni( rd );
    m_morning = ( random_hour < 12 );
}

//! \fn     SixStepTemplateMethod::setup
//! \brief  Base class setup method
//! \note   Not an overridden step
void SixStepTemplateMethod::setup()
{
    std::cout << std::endl << m_implementation_name << std::endl;
    std::cout << "setup" << std::endl;
}

//! \fn     SixStepTemplateMethod::schedule
//! \brief  Base class schedule method
//! \note   Not an overridden step
void SixStepTemplateMethod::schedule()
{
    std::cout << "schedule" << std::endl;
}

//! \fn     SixStepTemplateMethod::heat
//! \brief  Base class heat method
//! \note   Not an overridden step
void SixStepTemplateMethod::heat()
{
    std::cout
    << ( m_morning ? "high heat" : "low heat" )
    << std::endl;
}

//! \fn     SixStepTemplateMethod::optimize
//! \brief  Base class optimize method
//! \note   Default implementation of step that can be overridden
void SixStepTemplateMethod::optimize()
{
    std::cout << "standard optimization" << std::endl;
}

//! \fn     SixStepTemplateMethod::cleanup
//! \brief  Base class cleanup method
//! \note   Default implementation of step that can be overridden
void SixStepTemplateMethod::cleanup()
{
    std::cout << "standard cleanup" << std::endl;
}

//! \fn     SixStepTemplateMethod::putaway
//! \brief  Base class putaway method
//! \note   Not an overridden step
void SixStepTemplateMethod::putaway()
{
    std::cout << "putaway" << std::endl;
}

/*---------------------------------------------------------------------------*\
|   SixStepTemplateMethod template method
\*---------------------------------------------------------------------------*/
void SixStepTemplateMethod::template_method()
{
    setup();
    schedule();
    heat();
    optimize();
    cleanup();
    putaway();
}

/*---------------------------------------------------------------------------*\
|   SixStepTemplateMethodFast overrides
\*---------------------------------------------------------------------------*/
//! \fn     SixStepTemplateMethodFast::SixStepTemplateMethodFast
//! \brief  Disciple of the SixStepTemplateMethod
SixStepTemplateMethodFast::SixStepTemplateMethodFast(
    const char* name )
    : SixStepTemplateMethod( name )
{}

//! \fn     SixStepTemplateMethodFast::optimize
//! \brief  Overrides the optimize step
void SixStepTemplateMethodFast::optimize()
{
    std::cout << "fast optimize" << std::endl;
}

//! \fn     SixStepTemplateMethodFast::cleanup
//! \brief  Overrides the default cleanup step for illustration
void SixStepTemplateMethodFast::cleanup()
{
    std::cout << "standard cleanup" << std::endl;
}

/*---------------------------------------------------------------------------*\
|   SixStepTemplateMethodEconomical overrides
\*---------------------------------------------------------------------------*/
//! \fn     SixStepTemplateMethodEconomical::SixStepTemplateMethodEconomical
//! \brief  Disciple of the SixStepTemplateMethod
SixStepTemplateMethodEconomical::SixStepTemplateMethodEconomical(
    const char* name )
    : SixStepTemplateMethod( name )
{}

//! \fn     SixStepTemplateMethodEconomical::optimize
//! \brief  Overrides the optimize step
void SixStepTemplateMethodEconomical::optimize()
{
    std::cout << "economical optimize" << std::endl;
}

//! \fn     SixStemTemplateMethodEconomical::cleanup
//! \brief  Overrides the cleanup step for illustration
//! \note   Does NOT override the default step
void SixStepTemplateMethodEconomical::cleanup()
{
    std::cout << "standard cleanup" << std::endl;
}

}//namespace template_method_common

File: lab02/ruby-template-method/six_step_template_method.rb
#*----------------------------------------------------------------------------*\
#  Ruby version of Template Method Pattern lab
#  @author Norm Evangelista
#*----------------------------------------------------------------------------*/

module TemplateMethodCommon

#*----------------------------------------------------------------------------*\
#  Base Template Method class
#  @note Contains hook methods (default implementations) for steps to be overridden
#*----------------------------------------------------------------------------*/
class SixStepTemplateMethod

   #*-------------------------------------------------------------------------*\
   #  Constructor, generates random hour value to set @morning
   #*-------------------------------------------------------------------------*/
   def initialize( name = "" )
      @implementation_name = name

      hour     = Random.rand(0..23)
      @morning = (hour < 12)
   end

   #*-------------------------------------------------------------------------*\
   #  Default setup, not overridden
   #*-------------------------------------------------------------------------*/
   def setup
      puts "\n#{@implementation_name}"
      puts "setup"
   end

   #*-------------------------------------------------------------------------*\
   #  Default schedule, not overridden
   #*-------------------------------------------------------------------------*/
   def schedule
      puts "schedule"
   end

   #*-------------------------------------------------------------------------*\
   #  Default heat, not overridden
   #*-------------------------------------------------------------------------*/
   def heat
      if @morning
         puts "high heat"
      else
         puts "low heat"
      end
   end

   #*-------------------------------------------------------------------------*\
   #  Hook method for optimize, overridden by subclasses
   #*-------------------------------------------------------------------------*/
   def optimize
      puts "standard optimization"
   end

   #*-------------------------------------------------------------------------*\
   #  Hook method for cleanup, overridden by subclasses
   #*-------------------------------------------------------------------------*/
   def cleanup
      puts "standard cleanup"
   end

   #*-------------------------------------------------------------------------*\
   #  Default putaway, not overridden
   #*-------------------------------------------------------------------------*/
   def putaway
      puts "putaway"
   end

   #*-------------------------------------------------------------------------*\
   #  The heart of the Template Method: the template method
   #*-------------------------------------------------------------------------*/
   def template_method

      setup
      schedule
      heat
      optimize # Step to be overridden
      cleanup  # Step to be overridden
      putaway

   end

end

#*----------------------------------------------------------------------------*\
#  Fast variation of SixStepTemplateMethod
#  @note Only overrides SixStepTemplateMethod hook methods
#*----------------------------------------------------------------------------*/
class SixStepTemplateMethodFast < SixStepTemplateMethod

   #*-------------------------------------------------------------------------*\
   #  Constructor
   #*-------------------------------------------------------------------------*/
   def initialize
      super( "Fast method" )
   end

   #*-------------------------------------------------------------------------*\
   #  Overrides optimize step
   #*-------------------------------------------------------------------------*/
   def optimize
      puts "fast optimize"
   end

   #*-------------------------------------------------------------------------*\
   #  Overrides cleanup step
   #  @note Not really necessary, since it's identical to the base hook method
   #*-------------------------------------------------------------------------*/
   def cleanup
      puts "standard cleanup"
   end

end

#*----------------------------------------------------------------------------*\
#  Economical variation of SixStepTemplateMethod
#  @note Only overrides SixStepTemplateMethod hook methods
#*----------------------------------------------------------------------------*/
class SixStepTemplateMethodEconomical < SixStepTemplateMethod

   #*-------------------------------------------------------------------------*\
   #  Constructor
   #*-------------------------------------------------------------------------*/
   def initialize
      super( "Economical method" )
   end

   #*-------------------------------------------------------------------------*\
   #  Overrides optimize step
   #*-------------------------------------------------------------------------*/
   def optimize
      puts "economical optimize"
   end

   #*-------------------------------------------------------------------------*\
   #  Overrides cleanup step
   #  @note Not really necessary, since it's identical to the base hook method
   #*-------------------------------------------------------------------------*/
   def cleanup
      puts "standard cleanup"
   end

end

end #module TemplateMethodCommon
